Add Membership whitelist customization docs to console customization page

benjaminapetersen opened the pull request:
Per [this bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1577568)

Adds a short block that describes how the default list of roles listed on the membership page can be customized.  At present, the membership dropdown will list the following:

- `ClusterRole`s in this set: `  "admin",  "basic-user",  "edit",  "system:deployer",  "system:image-builder",  "system:image-puller",  "system:image-pusher",  "view",`
- any custom `Role` in a namespace/project 

This is a fairly sensible default as users will likely only need the specified `ClusterRole`s  above.  The additional `ClusterRole`s are overwhelming (but still visible via the "show all roles" checkbox). 

We also assume that any custom `Role` in a project is created for a reason, so ought to be displayed by default.

@ahardin-rh @adellape @spadgett

ahardin-rh commented on install_config/web_console_customization.adoc:
```diff
@@ -855,6 +855,32 @@ xref:../install_config/configuring_authentication.adoc#OpenID[OpenID] identity
+[[changing-the-membership-whitelist]]
+== Changing the Membership Whitelist
+
+The default whitelist in the membership page will show a certain subset of Cluster Roles
```
The default whitelist in the membership page shows a subset of cluster roles, such as `admin`, basic-user`, `edit`, and so on. It also shows custom roles defined within a project.

ahardin-rh commented on install_config/web_console_customization.adoc:
```diff
@@ -855,6 +855,32 @@ xref:../install_config/configuring_authentication.adoc#OpenID[OpenID] identity
+a project.
+
+For example, to add your own set of custom Cluster Roles to the whitelist:
+
```
s/Cluster Roles/cluster roles

yapei replied:
Is it should be `add your own set of custom roles to the whitelist`? Since we support both `Cluster Roles` and `Roles`(project/namespace scope)

benjaminapetersen replied:
You can add `ClusterRoles` and `Roles`, I suppose, but our filter already will approve all `Roles`, so there is no reason to add them.  It might read a little funny, but I wanted to be specific `cluster roles` to help imply this point.  We assume that a new `Role` is created for a pretty important reason (since there are already so many roles to choose from) so should show by default.

ahardin-rh commented on install_config/web_console_customization.adoc:
```diff
@@ -855,6 +855,32 @@ xref:../install_config/configuring_authentication.adoc#OpenID[OpenID] identity
+],
+----
+
+
```
Remove extra line

spadgett replied:
```
 Add the script as described in
 xref:loading-custom-scripts-and-stylesheets[Loading Extension Scripts and Stylesheets].
```

ahardin-rh commented on install_config/web_console_customization.adoc:
```diff
@@ -855,6 +855,32 @@ xref:../install_config/configuring_authentication.adoc#OpenID[OpenID] identity
+
+The default whitelist in the membership page will show a certain subset of Cluster Roles
+(such as 'admin', 'basic-user', 'edit', etc), as well as any custom roles defined within
+a project.
```
Why would someone want to change the membership whitelist? What's the use case for this? We should make that clear in the section intro.

benjaminapetersen replied:
From the bugzilla above, this person seems to want to show ONLY custom roles, so no cluster roles.  I'd guess they would `MEMBERSHIP_WHITELIST: [ ]` to eliminate ALL cluster roles.  Perhaps they basically recreated the main set of `basic-user`, `view`, etc, but have some subtle nuance for each.  

I'm guessing here and betting there would be some pain points with this approach.

ahardin-rh commented:
@benjaminapetersen I made some suggestions. Is this for 3.7+?

ahardin-rh commented:
@yapei Can you please give a QE review as well? Thanks!

yapei commented:
> At present, the membership dropdown will list the following:

    ClusterRoles in this set: "admin", "basic-user", "edit", "system:deployer", "system:image-builder", "system:image-puller", "system:image-pusher", "view",

Could we overwrite default `ClusterRoles` in above list by setting a new list in `MEMBERSHIP_WHITELIST`? Or are these hard-coded to list on Membership page?

spadgett commented on install_config/web_console_customization.adoc:
```diff
@@ -855,6 +855,32 @@ xref:../install_config/configuring_authentication.adoc#OpenID[OpenID] identity
 providers, which require visiting an external URL to destroy single sign-on
 sessions.
 
+[[changing-the-membership-whitelist]]
```
We should move this section up higher in the topic to the where the other extensions are described.

spadgett commented:
This should be backported to 3.7 docs. The way you add extensions in 3.7 is different, so it might need some small changes.

benjaminapetersen commented:
@yapei you could overwrite the list.  If you set `MEMBERSHIP_WHITELIST: []`, nothing would appear except namespaced roles.

benjaminapetersen commented:
Made the suggested changes & moved the block closer to the top.

ahardin-rh commented:
@benjaminapetersen Thanks! Can you please squash the changes into 1 commit? 

I will apply this change to 3.9+. It sounds like the process is a bit different for 3.7, so we'll need a separate PR against 3.7.

benjaminapetersen commented:
Yup, squashed.

benjaminapetersen commented:
For reference, the difference between the current filter && `3.7` is that `Roles` automatically pass the current filter:

```JavaScript
// Current
var filterRoles = function(roles) {
  return _.filter(roles, function(role) {
    return role.kind === 'Role' || _.includes(Constants.MEMBERSHIP_WHITELIST, role.metadata.name);
  });
};
// 3.7
var filterRoles = function(roles) {
  return _.filter(roles, function(item) {
    return _.includes(Constants.MEMBERSHIP_WHITELIST, item.metadata.name);
  });
};
```
So the older docs will probably just need to remove the `Cluster` qualifier.  I'll make another PR.

benjaminapetersen commented:
Just realized there are not release tags on this repo.

@ahardin-rh I might need a little guidance on the correct way to port this to 3.7 😄

benjaminapetersen commented:
@xingxingxia I don't know that we need to explicitly cover [#9215 (comment)](https://github.com/openshift/openshift-docs/pull/9251#issuecomment-390665469), we wouldn't encourage this use case, it would not typically be useful.  Someone who wants to customize the page to that degree would probably figure it out, assuming they are a developer & understand arrays.  That said, I defer to anyone who feels strongly otherwise.

xingxingxia commented:
> As to function, I did a quick try but seems didn't get expected result per doc

Hmm, to get issue earlier clarified, I tried again in the env http://pastebin.test.redhat.com/599902 (you can ssh with known key), I tried https://rawgit.com/xingxingxia/v3-testfiles/customize-web/extensions/membership-whitelist.js and https://rawgit.com/xingxingxia/v3-testfiles/customize-web/extensions/membership-whitelist-empty.js separately, neither works, the list is still unchanged. @benjaminapetersen could you check? If my setting is wrong please tell; if the doc is not simply copy-paste-working, please update for user to easy copy-paste-work :)
Thank you

benjaminapetersen commented on install_config/web_console_customization.adoc:
```diff
@@ -230,6 +230,33 @@ width and height. The ideal height is *20px*.
+== Customizing the Membership Whitelist
+
+The default whitelist in the membership page shows a subset of cluster roles, such as 
+`admin`, basic-user`, `edit`, and so on. It also shows custom roles defined within a project.
```
fixed.

ahardin-rh commented:
@benjaminapetersen Sorry for the delay! To submit changes against 3.7, we can open a new PR against a feature branch off of the 3.7 branch.

yapei commented on install_config/web_console_customization.adoc:
```diff
@@ -230,6 +230,33 @@ width and height. The ideal height is *20px*.
+
+[source, javascript]
+----
+MEMBERSHIP_WHITELIST: [
```
Could we update it to be more clear then users could copy and configure more easily?
```
window.OPENSHIFT_CONSTANTS.MEMBERSHIP_WHITELIST = [
  "admin",
  "basic-user",
  "edit",
  "system:deployer",
  "system:image-builder",
  "system:image-puller",
  "system:image-pusher",
  "view",
  "custom-role-1",
  "custom-role-2"
];
```

yapei replied:
Testing has passed per [bug comment 8](https://bugzilla.redhat.com/show_bug.cgi?id=1577568#c8) and [bug comment 9](https://bugzilla.redhat.com/show_bug.cgi?id=1577568#c9)

benjaminapetersen commented:
@ahardin-rh ok!  I'll let this one settle out on changes, then open that PR.

xingxingxia commented:
> Could we update it to be more clear then users could copy and configure more easily?

@benjaminapetersen when free, could you PTAL at yapei's https://github.com/openshift/openshift-docs/pull/9251#discussion_r193297113 and update it? So that users would not fail like me in https://github.com/openshift/openshift-docs/pull/9251#issuecomment-394721632

benjaminapetersen commented:
@xingxingxia @yapei sure thing!  Thanks.

xingxingxia commented on install_config/web_console_customization.adoc:
```diff
@@ -230,6 +230,33 @@ width and height. The ideal height is *20px*.
+
+[source, javascript]
+----
+window.OPENSHIFT_CONSTANTS.MEMBERSHIP_WHITELIST: [
```
Hmm, we tried like yapei's https://github.com/openshift/openshift-docs/pull/9251#discussion_r193297113, `:` does not work, only `=` works, please help fix it. With this fixed, the doc will be fine

benjaminapetersen commented:
@xingxingxia my apologies, I should have reviewed the whole code block when I added the `window. OPENSHIFT_CONSTANTS.` per your request.  Updated, I think this should work copy/paste w/o a problem.

xingxingxia commented:
@ahardin-rh the PR is fine now, please help merge if no other objection, thanks

benjaminapetersen commented:
Just a ping as I'm reviewing bugzilla, see this one still open.

vikram-redhat commented:
@ahardin-rh - please merge and move this along. Please make sure to CP to the 3.11 branch.

ahardin-rh commented:
/cherrypick enterprise-3.9

ahardin-rh commented:
/cherrypick enterprise-3.10

ahardin-rh commented:
/cherrypick enterprise-3.11
